Admin/BreadcrumbsBuilder.php

```
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+use Knp\Menu\ItemInterface as MenuItemInterface;
+
+/**
+ * Stateless breadcrumbs builder (each method needs an Admin object).
+ *
+ * @author Grégoire Paris <postmaster@greg0ire.fr>
+ */
+final class BreadcrumbsBuilder implements BreadcrumbsBuilderInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getBreadcrumbs(AdminInterface $admin, $action)
+    {
+        $breadcrumbs = array();
+        if ($admin->isChild()) {
+            return $admin->getParent()->getBreadcrumbs($action);
+        }
+
+        $menu = $admin->buildBreadcrumbs($action);
+
+        do {
+            $breadcrumbs[] = $menu;
+        } while ($menu = $menu->getParent());
```

Not sure what you mean by above : 
- above being my comment, this mean we should throw an exception earlier?
- above being your code snippet, what is the point of checking this if an exception is thrown before? The code will never be reached if there was an exception…